### ISSUE TITLE
html5ever is much slower than MyHTML

### ISSUE THREAD

**Issue by DemiMarie** [servo/html5ever#286]
<https://lexborisov.github.io/benchmark-html-persers/> shows html5ever performing very badly against [MyHTML](https://github.com/lexborisov/myhtml), which furthermore has fewer bugs and ~~supports `document.write()`~~ html5ever supports `document.write()` in modern versions (thanks @jdm).

**Comment by jdm** [comment]
html5ever also supports document.write().

**Comment by DemiMarie** [comment]
@jdm That must be an old version I was thinking of, sorry.

**Comment by SimonSapin** [comment]
Sorry, I don’t see the point of this issue. “Fewer bugs” is very non-specific. If you’re aware of bugs in the latest version that aren’t listed in https://github.com/servo/html5ever/issues, it’d be more helpful to file them individually.

As to performance, just stating “very bad” isn’t particularly actionable either. Are you interested in investigating low-hanging fruit to fix performance issues?

**Comment by nox** [comment]
That benchmark doesn't show how `html5ever` was compiled though.

**Comment by SimonSapin** [comment]
Closing for lack of actionable items.

**Comment by jdm** [comment]
The actionable item is attempting to reproduce the results. The benchmark repository looks like it's in good shape for doing that.

**Comment by Ygg01** [comment]
@jdm Perhaps issue #45  #51 might be more appropriate? I did ask about setting up benchmarks in a way that would work best for all parsers, but got no response, so I thought there wasn't any interest.

AFAICT the https://github.com/lexborisov/benchmark-html-persers uses C bindings for html5ever. I thought those were abandoned more than a year ago.

**Comment by jdm** [comment]
https://github.com/swizard0/html2html_lib/blob/master/src/lib.rs is the Rust code that parses the document; the benchmark calls that function from C.

**Comment by SimonSapin** [comment]
The C bindings in this repository were removed a year ago. IIRC they had been broken since we started using string-cache (since string-cache itself doesn’t have a C API), which happened 3 years ago. Also these bindings were only for tokenization.

According to [its `README.md`](https://github.com/lexborisov/benchmark-html-persers/blob/master/README.md), this benchmark uses [swizard0/html2html_lib](https://github.com/swizard0/html2html_lib) which:

* Decodes a UTF-8 byte string
* Parses into an `RcDom` tree
* Optionally serializes to stdout (this is disabled in the benchmark)
* Drops the tree

The tree cannot be accessed from C. (Which might be good enough for a parser.)

@nox pointed out that MyHTML does [YOLO ownership](https://github.com/lexborisov/myhtml/blob/v4.0.1/include/myhtml/tree.h#L84-L87) of tree nodes. Whereas our reference counting probably has some overhead. If we take the time to reproduce this benchmark, we should try using a tree with arena-allocated nodes. (And when publishing results, include a date and version numbers…)

**Comment by ordian** [comment]
I was able to build the benchmarks. See [fork](https://github.com/ordian/benchmark-html-persers).
[Here](https://docs.google.com/spreadsheets/d/1W2wb4SNT4LQLLt9X_Y5fkl7Yly_3X6EaDv-yqqMAUhg/edit?usp=sharing) are my local results with top(ish) 11 sites (you can see the plots by clicking on "explore" in the right bottom corner).

**Comment by DemiMarie** [comment]
MyHTML’s tree design only makes sense with a GC or with an immutable DOM.
So I think that we should try with html5ever having an arena.

**Comment by DemiMarie** [comment]
MyHTML’s tree design only makes sense with a GC or with an immutable DOM.
So I think that we should try with html5ever having an arena.  Gumbo has a pluggable allocator, so we should give it an arena too.

Could it be that the DOM-manipulation methods in RcDom are not getting inlined?  That would explain the slowdown.

**Comment by SimonSapin** [comment]
The benchmark’s `extern fn parse_document` uses `TendrilSink::read_from` with a `&[u8]` slice, treating it as a stream of unknown length that is copied into 4 KB chunks. It should use `one` instead, to use the slice as a single chunk. (Even that does a extra copy, though I don’t know if that’s significant compared to parsing time.)

**Comment by SimonSapin** [comment]
I’ve added an arena-allocated tree, as an alternative to rcdom: https://github.com/servo/html5ever/pull/288. It doesn’t use `unsafe` (except in the arena library), and shouldn’t have more overhead than a tree based on raw pointers.

**Comment by ordian** [comment]
@SimonSapin I've added your example to [fork](https://github.com/ordian/benchmark-html-parsers). 
![time](https://user-images.githubusercontent.com/4211399/28499387-3a02cf7e-6fbd-11e7-94d1-3d9991de39cc.png)

**Comment by SimonSapin** [comment]
So the arena making parsing a lot faster in some cases and a lot slower in others. This seems forth investigating, but I likely won’t take time to do that in the near future.

**Comment by ordian** [comment]
> So the arena making parsing a lot faster in some cases and a lot slower in others. 

I've [tried](https://github.com/ordian/benchmark-html-parsers/commit/e35195764597d7dac704f27439c106d0aca83e64) to make the benchmarks more scientific. Now arena-allocated tree doesn't look slower than rcdom.

**Comment by SimonSapin** [comment]
Looks good, thanks. This shows that reference-counting does have overhead, but that doesn’t account for all the difference with MyHTML. The next step is probably profiling to find hot spots in html5ever.

**Comment by DemiMarie** [comment]
According to `perf` 18% of the time is spent in `tendril::push_char`.  That looks somewhat suspicious.

**Comment by DemiMarie** [comment]
And 7.88 % is spent in `core::fmt::write_char`.  That should not even be called by html5ever.

**Comment by DemiMarie** [comment]
Here is a rasterized flame graph:
![flamegraph](https://user-images.githubusercontent.com/6395399/28501515-32473732-6fab-11e7-8038-952e334ee10b.png)
Here is the original vectorized version (rename from .txt to .svg to view it – GitHub does not allow it because SVG files can and do contain JavaScript):
[flamegraph.txt](https://github.com/servo/html5ever/files/1168260/flamegraph.txt)

**Comment by SimonSapin** [comment]
The tokenizer does work with `char`s, effectively decoding UTF-8 and then re-encoding it. We should consider working on slices or `u8` bytes, instead. This should be equivalent since all characters that are significant for HTML syntax are in the ASCII range instead. However this is not a small change, it’d take a significant amount of work.

**Comment by Ygg01** [comment]
@SimonSapin So if I understand correctly, this means we should use more `ByteTendril`s instead of `StrTendril`s, correct? Or iterate over `StrTendril` as if they were `ByteTendril`?

**Comment by SimonSapin** [comment]
I mean only changing internal details of the tokenizer. All APIs should keep using `StrTendril`.

**Comment by Ygg01** [comment]
So, internally, we should use more `u8`. This means internally, we should use `u8` instead of  `char`, correct? 

Will this require changes to `Tendril` or do we just use `tendril.borrow()`?

**Comment by SimonSapin** [comment]
I’d use `str::as_bytes` though `Deref<Target=str> for StrTendril`.

Also look if `Tendril::push_char` is used with `char`s from another tendril. In those cases, try instead to keep start and end indices around and then use `Tendril::subtendril` to get a reference-counted view of the same buffer, without copying data.

**Comment by Ygg01** [comment]
Yes, one interesting problem is how to solve [`get_preprocesed_char`](https://github.com/servo/html5ever/blob/abd8e2274b2232abf02873f4dfd399b0cf209bb4/html5ever/src/tokenizer/mod.rs#L265-L272) in an elegant way? I mean, I could create a random mask but it probably won't be as elegant as that. One thing that could work is perhaps some kind of macro that at compile time, essentially turn into that expression into a table lookup.

Second, is how to grab say next 4 or ten bytes from `BufferQueue`, theoretically speaking, first two bytes could be `tendril1` and next two could be `tendril2`, no?

**Comment by Ygg01** [comment]
@SimonSapin Is there a particular reason a Tokenizer doesn't own the `BufferQueue` given to it? Does having them separate enable some sort of workflow that can't be replicated otherwise?

**Comment by nox** [comment]
@Ygg01 If the tokeniser owns the buffer queue, we can't implement document.write nor speculative parsing.

**Comment by Ygg01** [comment]
Thanks nox, that's what I feared :(

In order to implement document.write/speculative parsing, what are the limitations? For example, does `BufferQueue` need `push_front`, `push_back`?

I ask, because keeping indices of a tendril for a let's say `temp_buf`, leads to some really weird syntax. I.e. it requires you to pass `BufferQueue` the only time, you don't need it, i.e. when you actually pass a char like `\u{FFFD}`, because you need to push the captured indices and then the character given, while when memorizing indices you don't need access to `BufferQueue`.

Is there a better place to discuss html5ever and when?

**Comment by Ygg01** [comment]
@nox Do you have any thoughts on:

> Also look if `Tendril::push_char` is used with `char`s from another tendril. In those cases, try instead to keep start and end indices around and then use `Tendril::subtendril` to get a reference-counted view of the same buffer, without copying data.

Especially, how does one keep indicies into another tendril, when `document.write` could potentially change the tendril? Note: I am assuming the indices are being kept by the `Tokenizer`.

**Comment by nox** [comment]
I don't understand your concern, keep indices during processing, then update the passed tendril so that the processed input is stripped off of its beginning.

**Comment by Ygg01** [comment]
At the risk of sounding obtuse, what counts as "during processing".

Hm, perhaps a better question would be  - When would you convert indices into subtendrils? At end of each   `state::Data => loop {}`?
